File: src/Hand.java
import java.util.Arrays;
import java.util.Comparator;

public class Hand {
    private Card[] cards;
    private HandVal handValue;

    public Hand(Card[] cards) {
        this.cards = cards;
        this.handValue = calcHandValue();
    }

    // constructor that takes a string describing 5 cards,
    // each with two characters, separated by at least one space
    // eg, "D2 HA ST CK S2"
    public Hand(String handDesc) {
        String[] cardDescs = handDesc.split("\\s+");
        if (cardDescs.length != 5) {
            throw new IllegalArgumentException("Invalid hand description: " + handDesc);
        }
        cards = new Card[5];
        for (int i = 0; i < 5; i++) {
            cards[i] = new Card(cardDescs[i]);
        }
        this.handValue = calcHandValue();
    }

    private HandVal calcHandValue() {
        // Placeholder for the actual hand value calculation logic
        // This will involve checking for various hand combinations
        // such as pairs, straights, flushes, etc.

        // Sort the cards by rank
        Arrays.sort(cards, Comparator.comparing(Card::getRank));

        // 1. Check for straight
        boolean isStraight = true;
        for (int i = 0; i < cards.length - 1; i++) {
            if (cards[i].getRank().getValue() != cards[i + 1].getRank().getValue() - 1) {
                isStraight = false;
                break;
            }
        }
        // Special case for Ace-low straight (A-2-3-4-5)
        if (!isStraight && cards[0].getRank() == Rank.TWO && cards[1].getRank() == Rank.THREE &&
                cards[2].getRank() == Rank.FOUR && cards[3].getRank() == Rank.FIVE && cards[4].getRank() == Rank.ACE) {
            isStraight = true;
        }

        // 2. checks for same suit
        boolean isSameSuit = Arrays.stream(cards).allMatch(card -> card.getSuit() == cards[0].getSuit());

        // 3. Straight flush or royal flush
        if (isStraight && isSameSuit) {
            // check royal flush
            if (cards[0].getRank() == Rank.TEN && cards[1].getRank() == Rank.JACK &&
                    cards[2].getRank() == Rank.QUEEN && cards[3].getRank() == Rank.KING &&
                    cards[4].getRank() == Rank.ACE) {
                return HandVal.ROYAL_FLUSH;
            } else {
                return HandVal.STRAIGHT_FLUSH;
            }
        }

        // 4. check for four of a kind
        for (int i = 0; i < cards.length - 3; i++) {
            int cardValue = cards[i].getRank().getValue();
            if (cardValue == cards[i + 1].getRank().getValue() &&
                    cardValue == cards[i + 2].getRank().getValue() &&
                    cardValue == cards[i + 3].getRank().getValue()) {
                return HandVal.POKER;
            }
        }

        // 5. check for full house

        // check triple
        boolean isThreeOfAKind = false;
        int isThreeOfAKindValue = -1;
        for (int i = 0; i < cards.length - 2; i++) {
            int cardValue = cards[i].getRank().getValue();
            if (cardValue == cards[i + 1].getRank().getValue() &&
                    cardValue == cards[i + 2].getRank().getValue()) {
                isThreeOfAKind = true;
                isThreeOfAKindValue = cardValue;
                break;
            }
        }

        // check pair
        boolean isPair = false;
        int isPairValue = -1;
        for (int i = 0; i < cards.length - 1; i++) {
            if (cards[i].getRank().getValue() == cards[i + 1].getRank().getValue()) {
                isPair = true;
                isPairValue = cards[i].getRank().getValue();
                if (isPairValue != isThreeOfAKindValue) {
                    break;
                }
            }
        }

        if (isThreeOfAKind && isPair && isThreeOfAKindValue != isPairValue) {
            return HandVal.FULL_HOUSE;
        }

        if (isSameSuit) {
            return HandVal.FLUSH;
        }

        if (isStraight) {
            return HandVal.STRAIGHT;
        }

        if (isThreeOfAKind) {
            return HandVal.THREE_OF_A_KIND;
        }

        // check for two pairs
        for (int i = 0; i < cards.length - 1; i++) {
            int cardValue = cards[i].getRank().getValue();
            if (cardValue == cards[i + 1].getRank().getValue()) {
                for (int j = i + 2; j < cards.length - 1; j++) {
                    int cardValue2 = cards[j].getRank().getValue();
                    if (cardValue2 == cards[j + 1].getRank().getValue()) {
                        return HandVal.TWO_PAIRS;
                    }
                }
            }
        }

        if (isPair) {
            return HandVal.PAIR;
        }

        // 5. Determine the highest value hand
        return HandVal.HIGH_CARD;
    }

    public Card[] getCards() {
        return cards;
    }

    public HandVal getHandValue() {
        return handValue;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Card card : cards) {
            sb.append(card).append("\n"); // new line after each card
        }
        return sb.toString().trim();
    }
}

File: src/golang/poker/hand.go
package poker

import (
	"errors"
	"fmt"
	"sort"
	"strings"
)

// Hand represents a poker hand of five cards
type Hand struct {
	Cards     []Card
	HandValue HandValue
}

// NewHand creates a Hand from an array of Card
func NewHand(cards []Card) (Hand, error) {
	if len(cards) != 5 {
		return Hand{}, errors.New("a hand must contain exactly 5 cards")
	}
	hand := Hand{Cards: cards}
	hand.HandValue = hand.calcHandValue()
	return hand, nil
}

// NewHandFromString creates a Hand from a string describing 5 cards, e.g., "D2 HA ST CK S2"
func NewHandFromString(handDesc string) (Hand, error) {
	cardDescs := strings.Fields(handDesc)
	if len(cardDescs) != 5 {
		return Hand{}, fmt.Errorf("invalid hand description: %s", handDesc)
	}
	cards := make([]Card, 5)
	for i, cardDesc := range cardDescs {
		card, err := NewCardFromString(cardDesc)
		if err != nil {
			return Hand{}, err
		}
		cards[i] = card
	}
	return NewHand(cards)
}

func (h *Hand) calcHandValue() HandValue {
	// Sort cards by rank
	sort.Slice(h.Cards, func(i, j int) bool {
		return h.Cards[i].Rank < h.Cards[j].Rank
	})

	// Check for straight
	isStraight := true
	for i := 0; i < len(h.Cards)-1; i++ {
		if h.Cards[i].Rank != h.Cards[i+1].Rank-1 {
			isStraight = false
			break
		}
	}
	// Special case for Ace-low straight (A-2-3-4-5)
	if !isStraight && h.Cards[0].Rank == TWO && h.Cards[1].Rank == THREE &&
		h.Cards[2].Rank == FOUR && h.Cards[3].Rank == FIVE && h.Cards[4].Rank == ACE {
		isStraight = true
	}

	// Check for same suit (flush)
	isFlush := true
	for _, card := range h.Cards {
		if card.Suit != h.Cards[0].Suit {
			isFlush = false
			break
		}
	}

	// Straight flush or royal flush
	if isStraight && isFlush {
		if h.Cards[0].Rank == TEN && h.Cards[1].Rank == JACK &&
			h.Cards[2].Rank == QUEEN && h.Cards[3].Rank == KING &&
			h.Cards[4].Rank == ACE {
			return ROYAL_FLUSH
		}
		return STRAIGHT_FLUSH
	}

	// Check for four of a kind
	for i := 0; i < len(h.Cards)-3; i++ {
		if h.Cards[i].Rank == h.Cards[i+1].Rank &&
			h.Cards[i].Rank == h.Cards[i+2].Rank &&
			h.Cards[i].Rank == h.Cards[i+3].Rank {
			return FOUR_OF_A_KIND
		}
	}

	// Check for full house
	isThreeOfAKind := false
	threeOfAKindRank := -1
	isPair := false
	pairRank := -1

	for i := 0; i < len(h.Cards)-2; i++ {
		if h.Cards[i].Rank == h.Cards[i+1].Rank &&
			h.Cards[i].Rank == h.Cards[i+2].Rank {
			isThreeOfAKind = true
			threeOfAKindRank = int(h.Cards[i].Rank)
			break
		}
	}

	for i := 0; i < len(h.Cards)-1; i++ {
		if h.Cards[i].Rank == h.Cards[i+1].Rank {
			isPair = true
			pairRank = int(h.Cards[i].Rank)
			if pairRank != threeOfAKindRank {
				break
			}
		}
	}

	if isThreeOfAKind && isPair && threeOfAKindRank != pairRank {
		return FULL_HOUSE
	}

	if isFlush {
		return FLUSH
	}

	if isStraight {
		return STRAIGHT
	}

	if isThreeOfAKind {
		return THREE_OF_A_KIND
	}

	// Check for two pairs
	pairCount := 0
	for i := 0; i < len(h.Cards)-1; i++ {
		if h.Cards[i].Rank == h.Cards[i+1].Rank {
			pairCount++
			i++ // skip next card as it's part of this pair
		}
	}
	if pairCount == 2 {
		return TWO_PAIR
	}

	if isPair {
		return PAIR
	}

	// Otherwise, it is a high card
	return HIGH_CARD
}

func (h Hand) String() string {
	var sb strings.Builder
	for _, card := range h.Cards {
		sb.WriteString(fmt.Sprintf("%v\n", card))
	}
	return strings.TrimSpace(sb.String())
}
